### PR TITLE
MTDSA-25058 -  fix for incorrect mapping for 403 from mtd-identifier-…

### PR DESCRIPTION
**Pull request by jeremystone**
…lookup

### CODE REVIEW

File: app/api/connectors/httpparsers/MtdIdLookupHttpParser.scala [hunk 1]
```diff
@@ -16,26 +16,20 @@
...
+  implicit val mtdIdLookupHttpReads: HttpReads[connectors.MtdIdLookupConnector.Outcome] = (_: String, _: String, response: HttpResponse) => {
     response.status match {
-      case OK =>
-        response.validateJson[String](mtdIdJsonReads) match {
-          case Some(mtdId) => Right(mtdId)
-          case None        => Left(InternalError)
-        }
-      case FORBIDDEN    => Left(NinoFormatError)
-      case UNAUTHORIZED => Left(InvalidBearerTokenError)
-      case _            => Left(InternalError)
```
Comment (Vedant-N421) [inline app/api/connectors/httpparsers/MtdIdLookupHttpParser.scala:38]: Does this mean that now the old 500s get mapped differently?

File: app/api/connectors/package.scala [hunk 2]
```diff
@@ -16,13 +16,11 @@
 
 package api
 
-import api.models.errors.{DownstreamError, MtdError}
+import api.models.errors.DownstreamError
 import api.models.outcomes.ResponseWrapper
 
 package object connectors {
 
-  type MtdIdLookupOutcome = Either[MtdError, String]
```
Comment (Vedant-N421) [inline app/api/connectors/package.scala:24]: Why have we made this to remove the type and instead add it to the connector?

File: app/api/controllers/validators/resolvers/ResolveNino.scala [hunk 3]
```diff
@@ -21,17 +21,23 @@ import api.models.errors.{MtdError, NinoFormatError}
 import cats.data.Validated
 import cats.data.Validated.{Invalid, Valid}
 
-object ResolveNino extends Resolver[String, Nino] {
+object ResolveNino {
+
+  def apply(value: String): Validated[Seq[MtdError], Nino] =
```
Comment (Vedant-N421) [inline app/api/controllers/validators/resolvers/ResolveNino.scala:26]: Why do we bring out the code from the Nino object? Not sure why we can't keep it to this:
```scala
object ResolveNino {

  private val ninoRegex =
    ("^([ACEHJLMOPRSWXY][A-CEGHJ-NPR-TW-Z]|B[A-CEHJ-NPR-TW-Z]|G[ACEGHJ-NPR-TW-Z]|" +
      "[KT][A-CEGHJ-MPR-TW-Z]|N[A-CEGHJL-NPR-SW-Z]|Z[A-CEGHJ-NPR-TW-Y])[0-9]{6}[A-D ]?$").r

  def apply(value: String): Validated[Seq[MtdError], Nino] = {
    if (Nino.isValid(value) && ninoRegex.matches(value)) {
      Valid(Nino(value))
    }
    else {
      Invalid(List(NinoFormatError))
    }
  }

}
```

File: app/api/services/EnrolmentsAuthService.scala [hunk 4]
```diff
@@ -39,55 +38,51 @@ class EnrolmentsAuthService @Inject() (val connector: AuthConnector, val appConf
...
+        case _: AuthorisationException => Future.successful(Left(ClientOrAgentNotAuthorisedError))
+        case error =>
+          logger.warn(s"[EnrolmentsAuthService][authorised] An unexpected error occurred: $error")
+          Future.successful(Left(InternalError))
+      }
   }
 
-  private def retrieveAgentDetails()(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[String]] =
+  private def retrieveAgentDetails()(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[String]] = {
+    def getAgentReferenceFrom(enrolments: Enrolments) =
```
Comment (Vedant-N421) [inline app/api/services/EnrolmentsAuthService.scala:75]: I think we should add the return type for the method as `Option[String]` to make it clearer?

File: app/api/connectors/httpparsers/MtdIdLookupHttpParser.scala (hunk 1 above)
Comment (jeremystone) [reply app/api/connectors/httpparsers/MtdIdLookupHttpParser.scala:38]: No. This code is essentially moved to the `MtdIdLookupService`. This allows the http parser & connector to not be concerned about the interpretation of errors coming from downstream. They just honestly report them to the service: 'this is what the lookup service returned'.

File: app/api/connectors/package.scala (hunk 2 above)
Comment (jeremystone) [reply app/api/connectors/package.scala:24]: Because the return type is tightly associated with the connector (and it now wraps a specific Error case class holding the status code rather than an MtdError).

File: app/api/services/EnrolmentsAuthService.scala (hunk 4 above)
Comment (jeremystone) [reply app/api/services/EnrolmentsAuthService.scala:75]: Hmm. Maybe. But it's a local method inside a private method and this is shared code so should be changed in BSAS if you feel strongly about it (otherwise it will get overwritten).

File: app/api/controllers/validators/resolvers/ResolveNino.scala (hunk 3 above)
Comment (jeremystone) [reply app/api/controllers/validators/resolvers/ResolveNino.scala:26]: But your wider point about the actual Nino object is a good one. There was a conversation started earlier today about why we do this extra validation in ResolveNino. No resolution yet, but hopefully it will get sorted in BSAS and rolled out if there is any change required.

File: app/utils/ErrorHandler.scala [hunk 5]
```diff
@@ -91,9 +99,9 @@ class ErrorHandler @Inject() (config: Configuration, auditConnector: AuditConnec
 
     val (errorCode, eventType) = ex match {
       case _: NotFoundException      => (NotFoundError, "ResourceNotFound")
-      case _: AuthorisationException => (ClientNotAuthenticatedError, "ClientError")
+      case _: AuthorisationException => (ClientOrAgentNotAuthorisedError.withStatus401, "ClientError")
       case _: JsValidationException  => (BadRequestError, "ServerValidationError")
-      case _: HttpException          => (BadRequestError, "ServerValidationError")
+      case e: HttpException          => (BadRequestError, "ServerValidationError")
```
Comment (Vedant-N421) [inline app/utils/ErrorHandler.scala:104]: This `e` is unused; revert this change?

File: app/api/controllers/validators/resolvers/ResolveNino.scala (hunk 3 above)
Comment (Vedant-N421) [reply app/api/controllers/validators/resolvers/ResolveNino.scala:26]: I'm still confused so a call about this might be useful

**Review by Vedant-N421 — APPROVED** [review]
